src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+section \<open>Chain Synchronization Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+subsection \<open>Cryptographic Primitives\<close>
+
+typedecl hash
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
+
+subsection \<open>Chains\<close>
+
+type_synonym chain = "block list"
+
+type_synonym chain_index = nat
+
+abbreviation (input) tip_index :: "chain \<Rightarrow> chain_index" where
+  "tip_index \<C> \<equiv> length \<C> - 1"
+
+abbreviation (input) tip :: "chain \<Rightarrow> block" where
+  "tip \<C> \<equiv> \<C> ! (tip_index \<C>)"
+
+subsection \<open>Points\<close>
+
+type_synonym point = hash
+
+definition block_point :: "block \<Rightarrow> point" where
+  [simp]: "block_point \<equiv> hash"
+
+subsection \<open>Protocol Parameters\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_chain_update: ChainUpdate
+
+type_synonym consumer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  phase \<comment> \<open>current phase\<close> \<times>
+  (chain \<Rightarrow> point list)" \<comment> \<open>function for selecting candidate points\<close>
+
+type_synonym producer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  chain_index \<comment> \<open>read pointer\<close> \<times>
+  bool \<comment> \<open>whether the next response to \<open>RequestNext\<close> should be \<open>RollBackward\<close>\<close>"
+
+locale chain_sync =
+  fixes initial_params\<^sub>c :: consumer_parameters
+    and initial_params\<^sub>p :: producer_parameters
```

Comment:
I think what you have covered here are not just protocol parameters, which should become parameters of the locale, but also internal state, which should only show up as arguments of programs. For example, the phase of the client is not a protocol parameter, as a user shouldn’t be able to specify in what phase the client starts.

In general, I somewhat like the bundling of protocol parameters according to whether they refer to the client or to the server. However, after dropping the parts that are not protocol parameters, there won’t be so many parameters left, so that it’ll probably be better to turn to a flat structure, similar to what has been done for the request-response protocol already. Also, other mini-protocols might have parameters that are not tied to specific parties, making the approach of grouping parameters according to party not universally applicable.

That said, whenever such bundling is done, it should not be done by introducing type synonyms for plain tuples. Instead, using records would be appropriate in such a situation. This would also free you from providing comments that say what certain elements are supposed to mean, because you would just name your record fields accordingly. This, in turn, would result in code that accesses elements not using nestings of `fst` and `snd` but the names that actually say what the selected items are about, meaning that your code would become much more readable and robust.

Once you’ve dropped internal state from this section, you can and should move the definition of the `phase` type into the subsection where the programs are defined, as the explicit representation of phases is really an implementation detail.

Also, I would not introduce a separate subsection *Protocol Parameters* here but rather have the specification of the protocol parameters directly belong to the enclosing section, like it is done with the ping-pong and the request-response protocol. I think this makes somehow sense, as this part is, well, about parameters, the things you feed in, and not about what this theory gives you. Not sure whether this makes very much sense, though. :sweat_smile: 

Finally, the alignment of the `fixes` and the `and` line in the `locale` specification will break when rendering this text using a proportional font, as it is done in document output (PDF files via LaTeX). I tend to use individual `fixes` specifications, like I did with the `request_response` locale. At least in the case of `assumes`, avoiding `and` this way helps with getting assumption names aligned.